sycl/source/detail/config.cpp

```diff
@@ -0,0 +1,104 @@
+//==---------------- config.cpp ---------------------------------*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "detail/config.hpp"
+
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <limits>
+
+#define STRINGIFY_LINE_HELP(s) #s
+#define STRINGIFY_LINE(s) STRINGIFY_LINE_HELP(s)
+
+namespace cl {
+namespace sycl {
+namespace detail {
+
+#ifndef SYCL_CONFIG_FILE_NAME
+#define SYCL_CONFIG_FILE_NAME "sycl.cfg"
+#endif // SYCL_CONFIG_FILE_NAME
+
+#define CONFIG(Name, MaxSize, CompileTimeDef)                                  \
+  const char *SYCLConfigBase<Name>::MValueFromFile = nullptr;                  \
+  char SYCLConfigBase<Name>::MStorage[MaxSize];                                \
+  const char *const SYCLConfigBase<Name>::MCompileTimeDef =                    \
+      getStrOrNullptr(STRINGIFY_LINE(CompileTimeDef));                         \
+  const char *const SYCLConfigBase<Name>::MConfigName = STRINGIFY_LINE(Name);
+#include "detail/config.def"
+#undef CONFIG
+
+#define MAX_CONFIG_NAME 256
+
+static void initValue(const char *Key, const char *Value) {
+#define CONFIG(Name, MaxSize, CompileTimeDef)                                  \
+  if (0 == strncmp(Key, SYCLConfigBase<Name>::MConfigName, MAX_CONFIG_NAME)) { \
+    strcpy(SYCLConfigBase<Name>::MStorage, Value);                             \
+    SYCLConfigBase<Name>::MValueFromFile = SYCLConfigBase<Name>::MStorage;     \
+    return;                                                                    \
+  }
+#include "detail/config.def"
+#undef CONFIG
+}
+
+void readConfig() {
+  static bool Initialized = false;
+  if (Initialized)
+    return;
+  std::fstream File;
+  // TODO: Find libsycl.so location.
+  static const char *ConfigFile = getenv("SYCL_CONFIG_FILE_NAME");
+  const char *FileToLoad = ConfigFile ? ConfigFile : SYCL_CONFIG_FILE_NAME;
+  File.open(FileToLoad, std::ios::in);
+  if (File.is_open()) {
+    // TODO: Use max size from macro instead of 256
+    char Key[MAX_CONFIG_NAME] = {0}, Value[256] = {0};
+    while (!File.eof()) {
+      // Expected fromat:
+      // ConfigName=Value
+      // ConfigName=Value
+      // TODO: Skip spaces before and after '='
+      File.getline(Key, sizeof(Key), '=');
+      if (File.fail()) {
+        // Fail to process the line. Skip it completely and try next one.
+        // Do we want to restore here? Or just throw an exception?
+        File.clear(File.rdstate() & ~std::ios_base::failbit);
+        File.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+        continue;
+      }
+      File.getline(Value, sizeof(Value), '\n');
```

Comment:
I wonder what happens on Windows. People tend add `\r\n` instead of `\n`.

Reply:
Added a logic which removes \r if it exists. Cannot write proper test as we have no env vars sensitive to the additional character at the end. Did manual testing.

Reply:
All this `\n` should be handled transparently by `iostream`, isn't it???

Reply:
Yes, \n is the default value of 3rd argument.. Will remove